embeddings guide + plot embeddings drift (#2503)
deepchecks/deepchecks: docs/source/nlp/usage_guides/nlp_embeddings.rst

.. _nlp__embeddings_guide:

=================
NLP Embeddings
=================

Embeddings are a way to represent text as a vector of numbers. The vector is a representation of the text in the latent

> space (or semantic space), in which text with similar meaning is represented by similar vectors.

Reviewer: Isn't semantic space only relevant when the model is trained on a semantic task? What if the model is trained to recognize language for e.g.?

Author: Fixed

> Embeddings are usually extracted from the last layer of a trained neural network model. This model can either be a

Reviewer: ```suggestion
Embeddings are usually extracted from the one of the final layers of a trained neural network model. This model can either be a
```

model that was trained on the specific task at hand (e.g. sentiment analysis), or a model that was trained on a

> different task, but is known to be good at extracting embeddings (e.g. BERT).

Reviewer: I would not use "BERT" as it's a kind of model that can be trained for a narrow task such as sentiment analysis. What about mentioning a specific pre-trained check point that was trained on a general task? Or even mention GPT, why not..

Author: I'm not sure I'm following why not BERT.
Also, would prefer not to mention something that is not open source such as GPT.

Reviewer: true about OSS
About BERT - my issue is that it's like saying "for embeddings you can either use a model trained for a specific task, or a multilayer perceptron". BERT is just an architecture, than is trained for a certain task. Like you are aiming at here - training some models on some tasks can result in embeddings that generalize, but that a result of the task (e.g. "predict the next word of the text") and the dataset (e.g. "a lot of the internet"). That's why I'm saying that as an example for that you don't want to point at an architecture, but rather at an example for a famous foundation model. An example for that are the BERT models trained as described in https://arxiv.org/abs/1908.10084.

What Are Embeddings Used For?
=============================

> Embeddings are used by some of the Deepchecks' checks, in order to extract meaningful

Reviewer: ```suggestion
Embeddings are used by some of the Deepchecks' checks to produce a meaningful representation of the data, 
```

    it may indicate that the model was trained on a dataset that is biased towards
    the news articles, and that the model is not generalizing well to the scientific articles.
#. **Find conflicting annotations** - Clean data is critical for training a good model. Mistakes in annotations
    (labeling) of the data can lead to a model that is not performing well. By finding similar texts (using embeddings)
    with different annotations, we can find potential annotation mistakes and fix them.

Using Embeddings in Checks
==========================

Whether you are :ref:`Using Deepchecks to Calculate Embeddings` or using your own model's embeddings, the process of
using them in the checks is the same.
In order to use the embeddings of your text in a check, the embeddings should already be part of the ``TextData`` object.

Using Deepchecks to Calculate Embeddings
----------------------------------------

If you don't have model embeddings for you text, you can use deepchecks to calculate the embeddings for you.
deepchecks currently supports using the free-to-use ``sentence-transformers`` library to calculate the embeddings,

> or the paid API of ``open-ai``.

Reviewer: ```suggestion
or the paid API of ``Open AI``.
```

Reviewer: mm wasn't clear to me

or Open AI's ``text-embedding-ada-002`` model. You can choose which model to use by setting the ``model`` parameter
to either ``miniLM`` or ``open_ai``.

The embeddings are automatically saved on a local CSV file so they can be used later. You can change the location and
name of the file by using the ``file_path`` parameter.

.. note:
    If you want to use the Open AI API, you will need to set the ``OPEN_AI_API_KEY`` environment variable to your
    Open AI API key. You can get your API key from the Open AI website.

Using Your Own Embeddings
-------------------------

Whether you saved the deepchecks embeddings for this dataset somewhere to save time, or you used your own model,
you can set the embeddings of the ``TextData`` object to be your own, by using one of the following methods:

#. When initializing the :class:`TextData <deepchecks.nlp.TextData>` object, pass your pre-calculated
   embeddings to the ``embeddings`` parameter.
#. After the initialization, call the ``set_embeddings`` method of the :class:`TextData <deepchecks.nlp.TextData>`
   object.

> In both methods, you can pass the embeddings as a pandas DataFrame, or as a path to a csv file. For the correct format

Reviewer: numpy!

of the embeddings, see the :ref:`Pre-Calculated Embeddings Format` section.

In the following example, we will pass pre-calculated embeddings to the ``TextData`` object in order to use the
TextPropertyOutliers check:

.. code-block:: python

  from deepchecks.nlp.checks import TextEmbeddingsDrift
  from deepchecks.nlp import TextData

  # Option 1: Initialize the TextData object with the embeddings:
  text_data = TextData(text, embeddings=embeddings)

  # Option 2: Initialize the TextData object and then set the embeddings:
  text_data = TextData(text)
  text_data.set_embeddings(embeddings)

  # Run the check
  TextEmbeddingsDrift().run(text_data)

Pre-Calculated Embeddings Format
################################

> The embeddings should be a pandas DataFrame, where each row represents a text sample and each column represents an

Author: No!
